data/pools/transactionPool.go

```diff
@@ -540,159 +584,143 @@ func (pool *TransactionPool) recomputeBlockEvaluator(committedTxIds map[transact
-			txgroupad[i].SignedTxn = tx
-		}
-		err := eval.TransactionGroup(txgroupad)
-		if err == ledger.ErrNoSpace {
-			stats.StopReason = telemetryspec.AssembleBlockFull
-			break
-		}
-		if err != nil {
-			// GOAL2-255: Don't warn for common case of txn already being in ledger
-			switch err.(type) {
-			case ledger.TransactionInLedgerError:
-				stats.CommittedCount++
-			case transactions.MinFeeError:
-				stats.InvalidCount++
-				logging.Base().Infof("Cannot add pending transaction to block: %v", err)
-			default:
-				stats.InvalidCount++
-				logging.Base().Warnf("Cannot add pending transaction to block: %v", err)
-			}
-		}
-	}
+			payset := assembled.Block().Payset
+			if len(payset) != 0 {
+				totalFees := uint64(0)
 
-	lvb, err := eval.GenerateBlock()
-	if err != nil {
-		return nil, fmt.Errorf("could not make proposals at round %d: could not finish evaluator: %v", round, err)
-	}
+				for i, txib := range payset {
+					fee := txib.Txn.Fee.Raw
+					encodedLen := len(protocol.Encode(&txib))
 
-	// Measure time here because we want to know how close to deadline we are
-	dt := time.Now().Sub(start)
-	stats.Nanoseconds = dt.Nanoseconds()
+					stats.IncludedCount++
+					totalFees += fee
 
-	if pool.logAssembleStats {
-		payset := lvb.Block().Payset
-		if len(payset) != 0 {
-			totalFees := uint64(0)
-
-			for i, txib := range payset {
-				fee := txib.Txn.Fee.Raw
-				encodedLen := len(protocol.Encode(&txib))
-
-				stats.IncludedCount++
-				totalFees += fee
-
-				if i == 0 {
-					stats.MinFee = fee
-					stats.MaxFee = fee
-					stats.MinLength = encodedLen
-					stats.MaxLength = encodedLen
-				} else {
-					if fee < stats.MinFee {
+					if i == 0 {
 						stats.MinFee = fee
-					} else if fee > stats.MaxFee {
 						stats.MaxFee = fee
-					}
-					if encodedLen < stats.MinLength {
 						stats.MinLength = encodedLen
-					} else if encodedLen > stats.MaxLength {
 						stats.MaxLength = encodedLen
+					} else {
+						if fee < stats.MinFee {
+							stats.MinFee = fee
+						} else if fee > stats.MaxFee {
+							stats.MaxFee = fee
+						}
+						if encodedLen < stats.MinLength {
+							stats.MinLength = encodedLen
+						} else if encodedLen > stats.MaxLength {
+							stats.MaxLength = encodedLen
+						}
 					}
+					stats.TotalLength += uint64(encodedLen)
 				}
-				stats.TotalLength += uint64(encodedLen)
+
+				stats.AverageFee = totalFees / uint64(stats.IncludedCount)
 			}
 
-			stats.AverageFee = totalFees / uint64(stats.IncludedCount)
-		}
+			var details struct {
+				Round uint64
+			}
+			details.Round = uint64(round)
+			logging.Base().Metrics(telemetryspec.Transaction, stats, details)
+		}()
+	}
 
-		var details struct {
-			Round uint64
-		}
-		details.Round = uint64(round)
-		logging.Base().Metrics(telemetryspec.Transaction, stats, details)
+	pool.assemblyMu.Lock()
+	defer pool.assemblyMu.Unlock()
+
+	pool.assemblyDeadline = deadline
+	deadline = deadline.Add(assemblyWaitEps)
```

Comment:
I'm not opposing to add 10ms to the 250ms delay, but could you explain why these 10ms are needed ?
i.e. isn't the transaction pool receives the OnNewBlock *always* before the agreement is attempting to make a proposal ?

Reply:
I wanted to avoid the scenario where `recomputeBlockEvaluator` broadcasts on the condition variable at 1ms post-deadline, but `AssembleBlock` wakes up at exactly the deadline and so fails to receive the block. The other approach I was thinking of was to have `recomputeBlockEvaluator` attempt to broadcast at some epsilon before the deadline.